packages/react/src/hooks/useTabs.ts

```diff
@@ -0,0 +1,229 @@
+import {useCallback, useState, useRef, type HTMLAttributes, type KeyboardEvent} from 'react'
+import {useId} from '@reach/auto-id'
+
+type TabOrientation = 'horizontal' | 'vertical'
+
+type UseTabsOptions = {
+  defaultTab?: string
+  autoActivate?: boolean
+  onTabActivate?: (id: string) => void
+  orientation?: TabOrientation
+}
+
+type TabState = {
+  activeTab: string | null
+  focusedTab: string | null
+  tabs: Set<string>
+}
+
+type TabListProps = {
+  label?: string
+  labelledBy?: string
+}
+
+type UseTabs = {
+  activeTab: string | null
+  focusedTab: string | null
+  activateTab: (id: string) => void
+  focusTab: (id: string) => void
+  getTabListProps: (props?: TabListProps) => HTMLAttributes<HTMLElement>
+  getTabProps: (id: string) => HTMLAttributes<HTMLElement>
+  getTabPanelProps: (id: string) => HTMLAttributes<HTMLElement>
+}
+
+export const useTabs = ({
+  defaultTab,
+  autoActivate = true,
+  orientation = 'horizontal',
+  onTabActivate,
+}: UseTabsOptions = {}): UseTabs => {
+  const uniqueId = useId()
+  const [state, _setState] = useState<TabState>({
+    activeTab: defaultTab || null,
+    focusedTab: null,
+    tabs: new Set(),
+  })
+
+  const setState = useCallback(
+    (updater: (state: TabState) => TabState) => {
+      _setState(prev => {
+        const nextState = updater(prev)
+
+        if (nextState.activeTab && nextState.activeTab !== prev.activeTab) {
+          onTabActivate?.(nextState.activeTab)
+        }
+
+        return nextState
+      })
+    },
+    [onTabActivate],
+  )
+
+  // Refs to store tab elements for focus management
+  const tabRefs = useRef<Map<string, HTMLElement>>(new Map())
+
+  const getTabId = useCallback((id: string) => `tabs-${uniqueId}-tab-${id}`, [uniqueId])
+  const getPanelId = useCallback((id: string) => `tabs-${uniqueId}-panel-${id}`, [uniqueId])
+
+  const registerTab = useCallback(
+    (id: string, element: HTMLElement) => {
+      if (tabRefs.current.has(id)) {
+        return
+      }
+
+      tabRefs.current.set(id, element)
+      setState(prev => ({
+        ...prev,
+        tabs: new Set([...prev.tabs, id]),
+        activeTab: prev.activeTab || id, // Set as active if no active tab
+      }))
+    },
+    [setState],
+  )
+
+  const focusTab = useCallback(
+    (id: string) => {
+      const element = tabRefs.current.get(id)
+      if (element) {
+        element.focus()
+        setState(prev => ({
+          ...prev,
+          focusedTab: id,
+        }))
+      }
+    },
+    [setState],
+  )
+
+  const activateTab = useCallback(
+    (id: string) => {
+      setState(prev => ({
+        ...prev,
+        activeTab: id,
+        focusedTab: id,
+      }))
+    },
+    [setState],
+  )
+
+  const getNextTabIdx = useCallback((idx: number, tabs: string[]) => (idx + 1) % tabs.length, [])
```

Comment:
Nit.. could we avoid abbreviation on the variable name please? Fine to keep it for argument name if you want though. Just for readability.

```suggestion
  const getNextTabIndex = useCallback((idx: number, tabs: string[]) => (idx + 1) % tabs.length, [])
```

Reply:
No prob 👍

bb3ef8d10174b4ace33faa1944c4a2283cf19aa0